core/src/test/java/io/undertow/server/handlers/GracefulShutdownHttp2TestCase.java

```diff
@@ -0,0 +1,196 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2021 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package io.undertow.server.handlers;
+
+import io.undertow.Handlers;
+import io.undertow.Undertow;
+import io.undertow.UndertowOptions;
+import io.undertow.client.ClientCallback;
+import io.undertow.client.ClientConnection;
+import io.undertow.client.ClientExchange;
+import io.undertow.client.ClientRequest;
+import io.undertow.client.UndertowClient;
+import io.undertow.protocols.http2.Http2Channel;
+import io.undertow.protocols.ssl.UndertowXnioSsl;
+import io.undertow.server.HttpHandler;
+import io.undertow.server.HttpServerExchange;
+import io.undertow.server.protocol.http2.Http2ServerConnection;
+import io.undertow.testutils.DefaultServer;
+import io.undertow.util.Headers;
+import io.undertow.util.Methods;
+import io.undertow.util.Protocols;
+
+import org.jboss.logging.Logger;
+import org.junit.AfterClass;
+import org.junit.Assert;
+import org.junit.BeforeClass;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.xnio.ChannelListener;
+import org.xnio.IoUtils;
+import org.xnio.OptionMap;
+import org.xnio.Options;
+import org.xnio.Xnio;
+import org.xnio.XnioWorker;
+
+import java.io.IOException;
+import java.net.URI;
+import java.util.Collections;
+import java.util.Set;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+@RunWith(DefaultServer.class)
+public class GracefulShutdownHttp2TestCase {
+    private static final Logger log = Logger.getLogger(GracefulShutdownHttp2TestCase.class);
+
+    private static Undertow server;
+    private static GracefulShutdownHandler shutdown;
+    private static int port = DefaultServer.getHostPort("default") + 1;
+
+    private static XnioWorker worker;
+
+    private static final Set<Http2Channel> activeHttp2Channels = Collections.newSetFromMap(new ConcurrentHashMap<>());
+    private static final AtomicReference<CountDownLatch> connectionClosedLatch = new AtomicReference<>();
+    private static final AtomicReference<CountDownLatch> requestReceivedLatch = new AtomicReference<>();
+
+    @BeforeClass
+    public static void setup() throws Exception {
+
+        shutdown = Handlers.gracefulShutdown(new HttpHandler() {
```

Comment:
You can use @BeforeServerStarts and @AfterServerStops here

Reply:
Done